README.rst
========
Overview
========


Python wrapper for strategies originally written in Fortran.

* Free software: MIT license

Installation
============

To use this library the original Fortran code must be compiled:
https://github.com/Axelrod-Python/TourExec.

::

    pip install axelrod-fortran


Usage
=====

Running a match:

.. code-block:: python

   >>> import axelrod_fortran as axlf
   >>> import axelrod as axl
   >>> p1 = axlf.Player('k31r')
   >>> p2 = axlf.Player('k33r')
   >>> match = axl.Match((p1, p2), turns=5)
   >>> match.play()
   [(C, C), (C, C), (C, D), (C, D), (C, C)]

Contributing
============

Please see `CONTRIBUTING.rst` for details about installing for development and
running the test suite.


CHANGELOG.rst
Changelog
=========

v0.4.8 (2020-08-20)
-------------------

* Remove support for Python 3.5
* Add support for Python 3.7
* Add support for random seeding

v0.4.7 (2020-03-13)
-------------------

* Rename strategies to match latest axelrod names
* Remove tag showing K59R as implemented by axl.RevisedDowning

v0.4.6 (2019-01-01)
-------------------

* Tag K58R as implemented by Rowsam

v0.4.5 (2018-10-15)
-------------------

* Add automatic copying of shared library for multiple copies of players

v0.4.4 (2018-03-15)
-------------------

* Tag K66R as implemented by Mikkelson

v0.4.3 (2018-02-07)
-------------------

* Tag K47R as implemented by RichardHufford
* Tag K51R as implemented by Colbert
* Tag K64R as implemented by Yamanhi

v0.4.2 (2018-01-10)
-------------------

* Tag K35R as implemented by Getzler
* Tag K41R as implemented by Weiner
* Tag K60R as implemented by GraaskampKatzen
* Tag K68R as implemented by Leyvraz
* Tag K72R as implemented by White
* Tag K75R as implemented by Harrington
* Tag K83R as implemented by Black
* Tag K84R as implemented by MoreTidemanAndChieruzz

v0.4.1 (2017-11-29)
-------------------

* Tag K31R as implemented by GoByMajority
* Tag K32R as implemented by Kluepfel
* Tag K42R as implemented by Borufsen
* Tag K44R as implemented by WmAdams
* Tag K49R as implemented by Cave

v0.4.0 (2017-10-20)
-------------------

* Add ability to pass name of shared library file

v0.3.2 (2017-10-13)
-------------------

* Tag K76R as implemented by axl.Gladstein
* Tag K67R as implemented by axl.Tranquilizer
* Tag K90R as implemented by axl.TitFor2Tats
* Tag K86R as implemented by axl.MoreGrofman
* Remove Tag showing K61R as implemented by axl.Champion

* Add test for Champion (K61R)

v0.3.1 (2017-08-04)
-------------------

* Identify strategy K61R as Champion by Danny Champion ranked at 2.

v0.3.0 (2017-08-04)
-------------------

* Identify strategy K92R as Tit For Tat by Anatol Rapoport ranked at 1.

v0.2.0 (2017-07-30)
-------------------

* Identify strategy K74R as participating in the second tournament, written
  by Edward Friedland and ranked at 61

`All changes between v0.1.0 and this release
<https://github.com/Axelrod-Python/axelrod-fortran/compare/v0.1.0...v0.2.0>`_

v0.1.0 (2017-07-29)
-------------------

* Player class to wrap a fortran strategy function into the axelrod library
* Characteristics dict with details of stochasticity, author and original
  ranking for each function
* Pre-built lists for all strategies and strategies which participated in
  Axelrod's second tournament


tests/test_player.py
from ctypes import c_int, c_float, POINTER
import itertools

import pytest

from axelrod_fortran import Player, characteristics, all_strategies
from axelrod import (Alternator, Cooperator, Defector, Match, MoranProcess,
                     Game, RandomGenerator, Tournament, basic_strategies)
from axelrod.action import Action


C, D = Action.C, Action.D


def test_init():
    for strategy in all_strategies:
        player = Player(strategy)
        is_stochastic = characteristics[strategy]['stochastic'] in (True, None)
        assert player.classifier['stochastic'] == is_stochastic
        assert player.original_name == strategy
        assert player.original_function.argtypes == (
            POINTER(c_int), POINTER(c_int), POINTER(c_int), POINTER(c_int),
            POINTER(c_float))
        assert player.original_function.restype == c_int
        with pytest.raises(ValueError):
            player = Player('test')


def test_matches():
    for strategy in all_strategies:
        for opponent in (Alternator, Cooperator, Defector):
            players = (Player(strategy), opponent())
            match = Match(players, turns=200)
            assert all(
                action in (C, D) for interaction in match.play()
                for action in interaction)


def test_noisy_matches():
    for strategy in all_strategies:
        for opponent in (Alternator, Cooperator, Defector):
            players = (Player(strategy), opponent())
            match = Match(players, turns=200, noise=0.5)
            assert all(
                action in (C, D) for interaction in match.play()
                for action in interaction)


def test_probend_matches():
    for strategy in all_strategies:
        for opponent in (Alternator, Cooperator, Defector):
            players = (Player(strategy), opponent())
            match = Match(players, prob_end=0.5, noise=0.5)
            assert all(
                action in (C, D) for interaction in match.play()
                for action in interaction)


def test_matches_with_different_game():
    for strategy in all_strategies:
        for opponent in (Alternator, Cooperator, Defector):
            game = Game(r=4, s=0, p=2, t=6)
            players = (Player(strategy), opponent())
            match = Match(players, turns=200, game=game)
            assert all(
                action in (C, D) for interaction in match.play()
                for action in interaction)


def test_original_strategy():
    """
    Test original strategy against all possible first 5 moves of a Match
    """
    actions_to_scores = {(0, 0): (3, 3), (0, 1): (0, 5),
                         (1, 0): (5, 0), (1, 1): (1, 1)}
    for strategy in all_strategies:
        for opponent_sequence in itertools.product((0, 1), repeat=5):

            player = Player(strategy)

            # Initial set up for empty history
            my_score, their_score = 0, 0
            move_number = 1
            their_previous_action, my_action = 0, 0

            for action in opponent_sequence:
                my_action = player.original_strategy(
                    their_last_move=their_previous_action,
                    move_number=move_number,
                    my_score=my_score,
                    their_score=their_score,
                    random_value=0,
                    my_last_move=my_action)

                assert my_action in [0, 1]

                scores = actions_to_scores[my_action, action]
                their_previous_action = action
                my_score += scores[0]
                their_score += scores[1]


def test_deterministic_strategies():
    """
    Test that the strategies classified as deterministic indeed act
    deterministically.
    """
    for strategy in all_strategies:
        player = Player(strategy)
        if player.classifier["stochastic"] is False:
            for opponent in basic_strategies:
                match = Match((player, opponent()))
                interactions = match.play()
                for _ in range(2):  # Repeat 3 matches
                    assert interactions == match.play(), player


def test_implemented_strategies():
    """
    Test that the deterministic strategies that are implemented in Axelrod
    give the same outcomes.
    """
    for strategy, dictionary in characteristics.items():
        axelrod_class = dictionary["axelrod-python_class"]
        stochastic = Player(strategy).classifier["stochastic"]
        if axelrod_class is not None and not stochastic:
            for opponent_strategy in basic_strategies:
                player = Player(strategy)
                opponent = opponent_strategy()
                match = Match((player, opponent))
                interactions = match.play()

                axl_player = axelrod_class()
                opponent = opponent_strategy()
                axl_match = Match((axl_player, opponent))
                axl_interactions = axl_match.play()
                assert interactions == axl_interactions


def test_champion_v_alternator():
    """
    Specific regression test for a bug. See:
    https://github.com/Axelrod-Python/axelrod-fortran/issues/62
    """
    player = Player("k61r")
    opponent = Alternator()
    seed = 3
    match = Match((player, opponent), seed=seed)
    interactions = match.play()
    assert interactions[25:30] == [(C, D), (C, C), (C, D), (D, C), (C, D)]
    match = Match((player, opponent), seed=seed)
    assert interactions == match.play()


def test_warning_for_self_interaction(recwarn):
    """
    Test that a warning is given for a self interaction.
    """
    player = Player("k42r")
    opponent = player
    match = Match((player, opponent))
    interactions = match.play()
    assert len(recwarn) == 1


def test_no_warning_for_normal_interaction(recwarn):
    """
    Test that a warning is not given for a normal interaction
    """
    for players in [(Player("k42r"), Alternator()),
                    (Player("k42r"), Player("k41r"))]:

        match = Match(players)
        interactions = match.play()
        assert len(recwarn) == 0


def test_multiple_copies(recwarn):
    players = [Player('ktitfortatc') for _ in range(5)] + [
        Player('k42r') for _ in range(5)]
    mp = MoranProcess(players)
    mp.play()
    mp.populations_plot()


def test_match_reproducibility():
    for _ in range(100):
        rng = RandomGenerator()
        seed = rng.random_seed_int()
        strategies = rng.choice(all_strategies, size=2)
        players1 = [Player(strategy) for strategy in strategies]
        match1 = Match(players1, turns=200, noise=0.1, seed=seed)
        results1 = match1.play()

        players2 = [Player(strategy) for strategy in strategies]
        match2 = Match(players2, turns=200, noise=0.1, seed=seed)
        results2 = match2.play()

        assert (results1 == results2)


def test_tournament_reproducibility():
    rng = RandomGenerator()
    seed = rng.random_seed_int()
    strategies = rng.choice(all_strategies, size=10)
    players1 = [Player(strategy) for strategy in strategies]
    tournament1 = Tournament(players1, seed=seed, repetitions=2)
    results1 = tournament1.play(processes=2)

    players2 = [Player(strategy) for strategy in strategies]
    tournament2 = Tournament(players2, seed=seed, repetitions=2)
    results2 = tournament2.play(processes=2)

    assert (results1.ranked_names == results2.ranked_names)



tests/test_titfortat.py
from axelrod_fortran import Player
import axelrod as axl
from axelrod.action import Action

C, D = Action.C, Action.D


def test_versus_alternator():
    players = (Player('ktitfortatc'), axl.Alternator())
    match = axl.Match(players, 5)
    expected = [(C, C), (C, D), (D, C), (C, D), (D, C)]
    assert match.play() == expected


def test_versus_cooperator():
    players = (Player('ktitfortatc'), axl.Cooperator())
    match = axl.Match(players, 5)
    expected = [(C, C), (C, C), (C, C), (C, C), (C, C)]
    assert match.play() == expected


def test_versus_defector():
    players = (Player('ktitfortatc'), axl.Defector())
    match = axl.Match(players, 5)
    expected = [(C, D), (D, D), (D, D), (D, D), (D, D)]
    assert match.play() == expected


def test_versus_itself():
    players = (Player('ktitfortatc'), Player('ktitfortatc'))
    match = axl.Match(players, 5)
    expected = [(C, C), (C, C), (C, C), (C, C), (C, C)]
    assert match.play() == expected


tests/test_strategies.py
from axelrod_fortran import characteristics


def test_original_rank():
    ranks = [
        details['original_rank']
        for details in characteristics.values()
        if details['original_rank'] is not None]

    assert sorted(ranks) == list(range(1, 63 + 1))


src/axelrod_fortran/player.py
from ctypes import byref, c_float, c_int, POINTER
import warnings

import axelrod as axl
from axelrod.interaction_utils import compute_final_score
from axelrod.action import Action

from .strategies import characteristics
from .shared_library_manager import MultiprocessManager, load_library

C, D = Action.C, Action.D
actions = {0: C, 1: D}
original_actions = {C: 0, D: 1}


self_interaction_message = """
You are playing a match with the same player against itself. However
axelrod_fortran players share memory. You can initialise another instance of an
Axelrod_fortran player with player.clone().
"""


# Initialize a module-wide manager for loading copies of the shared library.
manager = MultiprocessManager()
manager.start()
shared_library_manager = manager.SharedLibraryManager("libstrategies.so")


class Player(axl.Player):

    classifier = {"stochastic": True}

    def __init__(self, original_name):
        """
        Parameters
        ----------
        original_name: str
            The name of the fortran function from the original axelrod
            tournament
        game: axelrod.Game
            A instance of an axelrod Game
        """
        super().__init__()
        # The order of the next 4 lines is important. We must first check that
        # the player name is valid, then grab a copy of the shared library,
        # and then setup the actual strategy function.
        self.original_name = original_name
        self.index, self.shared_library_filename = \
            shared_library_manager.get_filename_for_player(self.original_name)
        self.shared_library = load_library(self.shared_library_filename)
        self.original_function = self.original_name

        is_stochastic = characteristics[self.original_name]['stochastic']
        if is_stochastic is not None:
            self.classifier['stochastic'] = is_stochastic

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.__original_function = None

    @property
    def original_name(self):
        return self.__original_name

    @original_name.setter
    def original_name(self, key):
        if key in characteristics:
            self.__original_name = key
        else:
            raise ValueError('{} is not a valid Fortran function'.format(key))

    @property
    def original_function(self):
        return self.__original_function

    @original_function.setter
    def original_function(self, value):
        self.__original_function = self.shared_library[(value + '_').lower()]
        self.__original_function.argtypes = (
            POINTER(c_int), POINTER(c_int), POINTER(c_int), POINTER(c_int),
            POINTER(c_float))
        self.__original_function.restype = c_int

    def original_strategy(
        self, their_last_move, move_number, my_score, their_score, random_value,
        my_last_move
    ):
        args = (
            c_int(their_last_move), c_int(move_number), c_int(my_score),
            c_int(their_score), c_float(random_value), c_int(my_last_move))
        return self.original_function(*[byref(arg) for arg in args])

    def strategy(self, opponent):
        if self is opponent:
            warnings.warn(message=self_interaction_message)

        if not self.history:
            their_last_move = 0
            scores = (0, 0)
            my_last_move = 0
        else:
            their_last_move = original_actions[opponent.history[-1]]
            scores = compute_final_score(zip(self.history, opponent.history),
                                         game=self.match_attributes["game"])
            my_last_move = original_actions[self.history[-1]]
        move_number = len(self.history) + 1
        if self.classifier["stochastic"]:
            random_value = self._random.random()
        else:
            random_value = 0
        original_action = self.original_strategy(
            their_last_move, move_number, scores[0], scores[1], random_value,
            my_last_move)
        return actions[original_action]

    def _release_shared_library(self):
        # While this looks like we're checking that the shared library file
        # isn't deleted, the exception is actually thrown if the manager
        # thread closes before the player class is garbage collected, which
        # tends to happen at the end of a script.
        try:
            name = self.original_name
            index = self.index
        except AttributeError:
            # If the Player does finish __init__, because the name of a
            # non-existent strategy is supplied, a copy of the shared library
            # won't be loaded, nor will self.original_name or self.index
            # exist. In that case there's nothing to do.
            return
        try:
            shared_library_manager.release(name, index)
        except FileNotFoundError:
            pass

    def reset(self):
        # Release the shared library since the object is rebuilt on reset.
        self._release_shared_library()
        super().reset()
        self.original_function = self.original_name

    def __del__(self):
        # Release the library before deletion.
        self._release_shared_library()

    def __repr__(self):
        return self.original_name


src/axelrod_fortran/shared_library_manager.py
from collections import defaultdict
from ctypes import cdll
from ctypes.util import find_library
from multiprocessing.managers import BaseManager
from pathlib import Path
import platform
import shutil
import subprocess
import tempfile
import uuid


def load_library(filename):
    """Loads a shared library."""
    lib = None
    if Path(filename).exists():
        lib = cdll.LoadLibrary(filename)
    return lib


class SharedLibraryManager(object):
    """LibraryManager creates (and deletes) copies of a shared library, which
    enables multiple copies of the same strategy to be run without the end user
    having to maintain many copies of the shared library.

    This works by making a copy of the shared library file and loading it into
    memory again. Loading the same file again will return a reference to the
    same memory addresses. To be thread-safe, this class just passes filenames
    back to the Player class (which actually loads a reference to the library),
    ensuring that multiple copies of a given player type do not use the same
    copy of the shared library.
    """

    def __init__(self, shared_library_name, verbose=False):
        self.shared_library_name = shared_library_name
        self.verbose = verbose
        self.filenames = []
        self.player_indices = defaultdict(set)
        self.player_next = defaultdict(set)
        # Generate a random prefix for tempfile generation
        self.prefix = str(uuid.uuid4())
        self.library_path = self.find_shared_library(shared_library_name)

    def find_shared_library(self, shared_library_name):
        # Hack for Linux since find_library doesn't return the full path.
        if 'Linux' in platform.system():
            output = subprocess.check_output(["ldconfig", "-p"])
            for line in str(output).split(r"\n"):
                rhs = line.split(" => ")[-1]
                if shared_library_name in rhs:
                    return rhs
            raise ValueError("{} not found".format(shared_library_name))
        else:
            return find_library(
                shared_library_name.replace("lib", "").replace(".so", ""))

    def create_library_copy(self):
        """Create a new copy of the shared library."""
        # Copy the library file to a new (temp) location.
        temp_directory = tempfile.gettempdir()
        copy_number = len(self.filenames)
        filename = "{}-{}-{}".format(
            self.prefix,
            str(copy_number),
            self.shared_library_name)
        new_filename = str(Path(temp_directory, filename))
        if self.verbose:
            print("Loading {}".format(new_filename))
        shutil.copy2(self.library_path, new_filename)
        self.filenames.append(new_filename)

    def next_player_index(self, name):
        """Determine the index of the next free shared library copy to
        allocate for the player. If none is available then make another copy."""
        # Is there a free index?
        if len(self.player_next[name]) > 0:
            return self.player_next[name].pop()
        # Do we need to load a new copy?
        player_count = len(self.player_indices[name])
        if player_count == len(self.filenames):
            self.create_library_copy()
            return player_count
        # Find the first unused index
        for i in range(len(self.filenames)):
            if i not in self.player_indices[name]:
                return i
        raise ValueError("We shouldn't be here.")

    def get_filename_for_player(self, name):
        """For a given player return a filename for a copy of the shared library
        for use in a Player class, along with an index for later releasing."""
        index = self.next_player_index(name)
        self.player_indices[name].add(index)
        if self.verbose:
            print("allocating {}".format(index))
        return index, self.filenames[index]

    def release(self, name, index):
        """Release the copy of the library so that it can be re-allocated."""
        self.player_indices[name].remove(index)
        if self.verbose:
            print("releasing {}".format(index))
        self.player_next[name].add(index)

    def __del__(self):
        """Cleanup temp files on object deletion."""
        for filename in self.filenames:
            path = Path(filename)
            if path.exists():
                if self.verbose:
                    print("deleting", str(path))
                path.unlink()


# Setup up thread safe library manager.
class MultiprocessManager(BaseManager):
    pass


MultiprocessManager.register('SharedLibraryManager', SharedLibraryManager)


src/axelrod_fortran/strategies.py
import axelrod as axl

characteristics = {
    'grasr': {
        'axelrod-python_class': None,
        'stochastic': True,
        'author': None,
        'original_rank': None},
    'k31r': {
        'axelrod-python_class': axl.GoByMajority,
        'stochastic': False,
        'author': 'Gail Grisell',
        'original_rank': 23},
    'k32r': {
        'axelrod-python_class': axl.SecondByKluepfel,
        'stochastic': True,
        'author': 'Charles Kluepfel',
        'original_rank': 10},
    'k33r': {
        'axelrod-python_class': None,
        'stochastic': False,
        'author': 'Harold Rabbie',
        'original_rank': 59},
    'k34r': {
        'axelrod-python_class': axl.Grudger,
        'stochastic': False,
        'author': 'James W Friedman',
        'original_rank': 52},
    'k35r': {
        'axelrod-python_class': axl.SecondByGetzler,
        'stochastic': True,
        'author': 'Abraham Getzler',
        'original_rank': 11},
    'k36r': {
        'axelrod-python_class': None,
        'stochastic': True,
        'author': 'Roger Hotz',
        'original_rank': 63},
    'k37r': {
        'axelrod-python_class': None,
        'stochastic': False,
        'author': 'George Lefevre',
        'original_rank': 37},
    'k38r': {
        'axelrod-python_class': None,
        'stochastic': False,
        'author': 'Nelson Weiderman',
        'original_rank': 34},
    'k39r': {
        'axelrod-python_class': None,
        'stochastic': False,
        'author': 'Tom Almy',
        'original_rank': 25},
    'k40r': {
        'axelrod-python_class': None,
        'stochastic': True,
        'author': 'Robert Adams',
        'original_rank': 35},
    'k41r': {
        'axelrod-python_class': axl.SecondByWeiner,
        'stochastic': False,
        'author': 'Herb Weiner',
        'original_rank': 7},
    'k42r': {
        'axelrod-python_class': axl.SecondByBorufsen,
        'stochastic': False,
        'author': 'Otto Borufsen',
        'original_rank': 3},
    'k43r': {
        'axelrod-python_class': None,
        'stochastic': False,
        'author': 'R D Anderson',
        'original_rank': 39},
    'k44r': {
        'axelrod-python_class': axl.SecondByWmAdams,
        'stochastic': True,
        'author': 'William Adams',
        'original_rank': 5},
    'k45r': {
        'axelrod-python_class': None,
        'stochastic': False,
        'author': 'Michael F McGurrin',
        'original_rank': 50},
    'k46r': {
        'axelrod-python_class': None,
        'stochastic': True,
        'author': 'Graham J Eatherley',
        'original_rank': 14},
    'k47r': {
        'axelrod-python_class': axl.SecondByRichardHufford,
        'stochastic': False,
        'author': 'Richard Hufford',
        'original_rank': 16},
    'k48r': {
        'axelrod-python_class': None,
        'stochastic': False,
        'author': 'George Hufford',
        'original_rank': 53},
    'k49r': {
        'axelrod-python_class': axl.SecondByCave,
        'stochastic': True,
        'author': 'Rob Cave',
        'original_rank': 4},
    'k50r': {
        'axelrod-python_class': None,
        'stochastic': True,
        'author': 'Rik Smoody',
        'original_rank': 54},
    'k51r': {
        'axelrod-python_class': axl.SecondByColbert,
        'stochastic': False,
        'author': 'John William Colbert',
        'original_rank': 18},
    'k52r': {
        'axelrod-python_class': None,
        'stochastic': True,
        'author': 'David A Smith',
        'original_rank': 48},
    'k53r': {
        'axelrod-python_class': None,
        'stochastic': True,
        'author': 'Henry Nussbacher',
        'original_rank': 45},
    'k54r': {
        'axelrod-python_class': None,
        'stochastic': False,
        'author': 'William H Robertson',
        'original_rank': 58},
    'k55r': {
        'axelrod-python_class': None,
        'stochastic': False,
        'author': 'Steve Newman',
        'original_rank': 42},
    'k56r': {
        'axelrod-python_class': None,
        'stochastic': False,
        'author': 'Stanley F Quayle',
        'original_rank': 38},
    'k57r': {
        'axelrod-python_class': None,
        'stochastic': False,
        'author': 'Rudy Nydegger',
        'original_rank': 31},
    'k58r': {
        'axelrod-python_class': axl.SecondByRowsam,
        'stochastic': False,
        'author': 'Glen Rowsam',
        'original_rank': 21},
    'k59r': {
        'axelrod-python_class': None,
        'stochastic': False,
        'author': 'Leslie Downing',
        'original_rank': 40},
    'k60r': {
        'axelrod-python_class': axl.SecondByGraaskampKatzen,
        'stochastic': False,
        'author': 'Jim Graaskamp and Ken Katzen',
        'original_rank': 6},
    'k61r': {
        'axelrod-python_class': None,
        'stochastic': True,
        'author': 'Danny C Champion',
        'original_rank': 2},
    'k62r': {
        'axelrod-python_class': None,
        'stochastic': True,
        'author': 'Howard R Hollander',
        'original_rank': 51},
    'k63r': {
        'axelrod-python_class': None,
        'stochastic': False,
        'author': 'George Duisman',
        'original_rank': 57},
    'k64r': {
        'axelrod-python_class': axl.SecondByYamachi,
        'stochastic': False,
        'author': 'Brian Yamachi',
        'original_rank': 17},
    'k65r': {
        'axelrod-python_class': None,
        'stochastic': False,
        'author': 'Mark F Batell',
        'original_rank': 47},
    'k66r': {
        'axelrod-python_class': axl.SecondByMikkelson,
        'stochastic': False,
        'author': 'Ray Mikkelson',
        'original_rank': 20},
    'k67r': {
        'axelrod-python_class': axl.SecondByTranquilizer,
        'stochastic': True,
        'author': 'Craig Feathers',
        'original_rank': 27},
    'k68r': {
        'axelrod-python_class': axl.SecondByLeyvraz,
        'stochastic': True,
        'author': 'Fransois Leyvraz',
        'original_rank': 12},
    'k69r': {
        'axelrod-python_class': None,
        'stochastic': True,
        'author': 'Johann Joss',
        'original_rank': 29},
    'k70r': {
        'axelrod-python_class': None,
        'stochastic': True,
        'author': 'Robert Pebly',
        'original_rank': 32},
    'k71r': {
        'axelrod-python_class': None,
        'stochastic': False,
        'author': 'James E Hall',
        'original_rank': 60},
    'k72r': {
        'axelrod-python_class': axl.SecondByWhite,
        'stochastic': True,
        'author': 'Edward C White Jr',
        'original_rank': 13},
    'k73r': {
        'axelrod-python_class': None,
        'stochastic': False,
        'author': 'George Zimmerman',
        'original_rank': 41},
    'k74r': {
        'axelrod-python_class': None,
        'stochastic': False,
        'author': 'Edward Friedland',
        'original_rank': 61},
    'k74rxx': {
        'axelrod-python_class': None,
        'stochastic': False,
        'author': 'Edward Friedland',
        'original_rank': None},
    'k75r': {
        'axelrod-python_class': axl.SecondByHarrington,
        'stochastic': True,
        'author': 'Paul D Harrington',
        'original_rank': 8},
    'k76r': {
        'axelrod-python_class': axl.SecondByGladstein,
        'stochastic': False,
        'author': 'David Gladstein',
        'original_rank': 46},
    'k77r': {
        'axelrod-python_class': None,
        'stochastic': True,
        'author': 'Scott Feld',
        'original_rank': 55},
    'k78r': {
        'axelrod-python_class': None,
        'stochastic': True,
        'author': 'Fred Mauk',
        'original_rank': 19},
    'k79r': {
        'axelrod-python_class': None,
        'stochastic': False,
        'author': 'Dennis Ambuehl and Kevin Hickey',
        'original_rank': 26},
    'k80r': {
        'axelrod-python_class': None,
        'stochastic': False,
        'author': 'Robyn M Dawes and Mark Batell',
        'original_rank': 36},
    'k81r': {
        'axelrod-python_class': None,
        'stochastic': False,
        'author': 'Martyn Jones',
        'original_rank': 43},
    'k82r': {
        'axelrod-python_class': None,
        'stochastic': True,
        'author': 'Robert A Leyland',
        'original_rank': 49},
    'k83r': {
        'axelrod-python_class': axl.SecondByBlack,
        'stochastic': True,
        'author': 'Paul E Black',
        'original_rank': 15},
    'k84r': {
        'axelrod-python_class': axl.SecondByTidemanAndChieruzzi,
        'stochastic': False,
        'author': 'T Nicolaus Tideman and Paula Chieruzz',
        'original_rank': 9},
    'k85r': {
        'axelrod-python_class': None,
        'stochastic': False,
        'author': 'Robert B Falk and James M Langsted',
        'original_rank': 33},
    'k86r': {
        'axelrod-python_class': axl.SecondByGrofman,
        'stochastic': False,
        'author': 'Bernard Grofman',
        'original_rank': 28},
    'k87r': {
        'axelrod-python_class': None,
        'stochastic': True,
        'author': 'E E H Schurmann',
        'original_rank': 44},
    'k88r': {
        'axelrod-python_class': None,
        'stochastic': True,
        'author': 'Scott Appold',
        'original_rank': 22},
    'k89r': {
        'axelrod-python_class': None,
        'stochastic': False,
        'author': 'Gene Snodgrass',
        'original_rank': 56},
    'k90r': {
        'axelrod-python_class': axl.TitFor2Tats,
        'stochastic': False,
        'author': 'John Maynard Smith',
        'original_rank': 24},
    'k91r': {
        'axelrod-python_class': None,
        'stochastic': False,
        'author': 'Jonathan Pinkley',
        'original_rank': 30},
    'k92r': {
        'axelrod-python_class': axl.TitForTat,
        'stochastic': False,
        'author': 'Anatol Rapoport',
        'original_rank': 1},
    'k93r': {
        'axelrod-python_class': None,
        'stochastic': True,
        'author': None,
        'original_rank': None},
    'kpavlovc': {
        'axelrod-python_class': axl.WinStayLoseShift,
        'stochastic': False,
        'author': None,
        'original_rank': None},
    'krandomc': {
        'axelrod-python_class': axl.Random,
        'stochastic': True,
        'author': None,
        'original_rank': 62},
    'ktf2tc': {
        'axelrod-python_class': axl.TitFor2Tats,
        'stochastic': False,
        'author': None,
        'original_rank': None},
    'ktitfortatc': {
        'axelrod-python_class': axl.TitForTat,
        'stochastic': False,
        'author': None,
        'original_rank': None},
}

all_strategies = list(characteristics.keys())

# Players from Axelrod's second tournament.
second_tournament_strategies = [
    name for name in characteristics.keys()
    if characteristics[name]["original_rank"] is not None]

representative_strategies = ['k60r', 'k91r', 'k40r', 'k76r', 'k67r']
